sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -250,33 +257,116 @@ private static DoFnSignature parseSignature(Class<? extends DoFn<?, ?>> fnClass)
+      Collection<MemberT> declaredMembersWithAnnotation(
+          Class<? extends Annotation> anno,
+          Class<?> startClass,
+          Class<?> stopClass,
+          MemberGetter<MemberT> getter) {
+    Collection<MemberT> matches = new ArrayList<>();
 
     Class<?> clazz = startClass;
     LinkedHashSet<Class<?>> interfaces = new LinkedHashSet<>();
 
     // First, find all declared methods on the startClass and parents (up to stopClass)
     while (clazz != null && !clazz.equals(stopClass)) {
-      for (Method method : clazz.getDeclaredMethods()) {
-        if (method.isAnnotationPresent(anno)) {
-          matches.add(method);
+      for (MemberT member : getter.getMembers(clazz)) {
+        if (member.isAnnotationPresent(anno)) {
+          matches.add(member);
         }
       }
 
-      Collections.addAll(interfaces, clazz.getInterfaces());
+      // Add all interfaces, including transitive
+      for (TypeDescriptor<?> iface : TypeDescriptor.of(clazz).getInterfaces()) {
+        interfaces.add(iface.getRawType());
+      }
 
       clazz = clazz.getSuperclass();
     }
 
     // Now, iterate over all the discovered interfaces
-    for (Method method : ReflectHelpers.getClosureOfMethodsOnInterfaces(interfaces)) {
-      if (method.isAnnotationPresent(anno)) {
-        matches.add(method);
+    for (Class<?> iface : interfaces) {
+      for (MemberT member : getter.getMembers(iface)) {
+        if (member.isAnnotationPresent(anno)) {
+          matches.add(member);
+        }
       }
     }
     return matches;
   }
 
+  private static ImmutableMap<String, DoFnSignature.StateDeclaration> analyzeStateDeclarations(
+      ErrorReporter errors,
+      Class<?> fnClazz) {
+
+    Map<String, DoFnSignature.StateDeclaration> declarations = new HashMap<>();
+
+    for (Field field : declaredFieldsWithAnnotation(DoFn.StateId.class, fnClazz, DoFn.class)) {
+      String id = field.getAnnotation(DoFn.StateId.class).value();
+
+      if (declarations.containsKey(id)) {
+        errors.throwIllegalArgument(
+            "Duplicate %s \"%s\", used on both of [%s] and [%s]",
+            DoFn.StateId.class.getSimpleName(),
+            id,
+            field.toString(),
+            declarations.get(id).field().toString());
+        continue;
+      }
+
+      Class<?> stateSpecRawType = field.getType();
+      if (!(stateSpecRawType.equals(StateSpec.class))) {
+        errors.throwIllegalArgument(
+                "%s annotation on non-%s field [%s] that has class %s",
+            DoFn.StateId.class.getSimpleName(),
+            StateSpec.class.getSimpleName(),
+            field.toString(),
+            stateSpecRawType.getName());
+        continue;
+      }
+
+      Type stateSpecType = field.getGenericType();
+
+      // By static typing this is already a well-formed State subclass
+      TypeDescriptor<?> stateType =
+          TypeDescriptor.of(((ParameterizedType) stateSpecType).getActualTypeArguments()[1]);
```

Comment:
I think you need to use TypeDescriptor.resolve() here somewhere. E.g. imagine you have a class `DoFnWithFooState<T> extends DoFn<String, String> { State<T> foo; }` and `MyFn extends DoFnWithFooState<Integer>` - it should resolve to Integer. Please add a test like that too.